RemoteIocServer/remote_ioc_server.py

```diff
@@ -0,0 +1,156 @@
+from __future__ import print_function, unicode_literals, division, absolute_import
+
+import argparse
+import sys
+import os
+import traceback
+import six
+
+from pcaspy import SimpleServer, Driver
+
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir)))
+from RemoteIocServer.config_monitor import ConfigurationMonitor
+from RemoteIocServer.gateway import GateWay
+from RemoteIocServer.pvdb import STATIC_PV_DATABASE, PvNames
+from RemoteIocServer.utilities import print_and_log, get_hostname_from_prefix
+from BlockServer.core.ioc_control import IocControl
+
+
+DEFAULT_GATEWAY_START_BAT = os.path.join("C:\\", "Instrument", "Apps", "EPICS", "gateway", "start_remoteioc_server.bat")
+
+
+def _error_handler(func):
+    @six.wraps(func)
+    def wrapper(*args, **kwargs):
+        try:
+            return func(*args, **kwargs)
+        except Exception:
+            print_and_log(traceback.format_exc())
+    return wrapper
+
+
+class RemoteIocListDriver(Driver):
+    def __init__(self, ioc_names, pv_prefix, gateway_pvlist_path, gateway_acf_path, gateway_restart_script_path):
+        super(RemoteIocListDriver, self).__init__()
+        self._remote_pv_prefix = None
+
+        self._ioc_controller = IocControl(pv_prefix)
+
+        self._iocs = ioc_names
+
+        self._gateway = GateWay(
+            local_pv_prefix=pv_prefix,
+            gateway_pvlist_file_path=gateway_pvlist_path,
+            gateway_acf_path=gateway_acf_path,
+            gateway_restart_script_path=gateway_restart_script_path
+        )
+        self._gateway.set_remote_pv_prefix(self._remote_pv_prefix)
+        self._gateway.set_ioc_list(ioc_names)
+
+        self._configuration_monitor = ConfigurationMonitor(
+            local_pv_prefix=pv_prefix,
+            restart_iocs_callback=self.restart_all_iocs
+        )
+
+        self.updatePVs()
+
+    @_error_handler
+    def write(self, reason, value):
+        print_and_log("RemoteIocListDriver: Processing PV write for reason {}".format(reason))
+        if reason == PvNames.INSTRUMENT:
+            self.set_remote_pv_prefix(value)
+        else:
+            print_and_log("RemoteIocListDriver: Could not write to PV '{}': not known".format(reason), "MAJOR")
```

Comment:
If the remote machine "get host name" doesn't work then you probably need to put the instrument PV into alarm. You could maybe also do with an error PV. Feel free to put this in a future ticket.